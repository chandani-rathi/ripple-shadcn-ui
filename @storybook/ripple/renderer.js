// renderer.js
export async function renderToCanvas({ storyFn, canvasElement }) {
  console.log("renderer")
  const output = storyFn();

  // Example: simple HTML render
  if (typeof output === 'string') {
    canvasElement.innerHTML = output;
  } else if (output instanceof HTMLElement) {
    canvasElement.innerHTML = '';
    canvasElement.appendChild(output);
  } else {
    canvasElement.innerHTML = `<pre>Unsupported output: ${JSON.stringify(output)}</pre>`;
  }
}
