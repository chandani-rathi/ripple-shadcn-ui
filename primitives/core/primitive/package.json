{
  "name": "@ripple-primitives/primitive",
  "version": "1.0.0",
  "description": "",
  "source": "./src/index.ts",
  "main": "./src/index.ts",
  "module": "./src/index.ts",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "packageManager": "pnpm@10.16.1"
}