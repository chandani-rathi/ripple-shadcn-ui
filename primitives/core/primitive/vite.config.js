import { defineConfig } from 'vite';
import { ripple } from 'vite-plugin-ripple';
import path from "node:path"

export default defineConfig({
	plugins: [ripple()],
	resolve: {
		alias: {
			'@': path.resolve(__dirname, './src'),
			'@ripple-primitives/primitive': path.resolve(__dirname, '../../core', 'primitive', 'src'),
			'@ripple-primitives/primitive-ui': path.resolve(__dirname, '../', 'primitive', 'src'),
		},
	},
	build: {
		target: 'esnext',
		manifest: false,
		minify: false,
		lib: {
			entry: 'src/index.ts',
			fileName: (format) => `index.${format}.js`,
			formats: ['es', 'cjs'],
		},
		rollupOptions: {
			input: 'src/index.ts',
			output: {
				dir: 'dist',
				globals: {},
				exports: 'named',
			},
			external: ['ripple', 'ripple/internal/client', '@ripple-primitives/primitive-ui', '@ripple-primitives/primitive'],
			plugins: [],
		},
	},
});
