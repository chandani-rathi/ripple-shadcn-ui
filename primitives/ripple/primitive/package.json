{
  "name": "@ripple-primitives/primitive-ui",
  "version": "1.0.0",
  "type": "module",
  "source": "./src/index.ts",
  "main": "./src/index.ts",
  "module": "./src/index.ts",
  "exports": {
    ".": {
      "import": "./dist/index.es.js",
      "require": "./dist/index.cjs.js"
    }
  },
  "scripts": {
    "build": "vite build",
    "watch": "vite build --watch",
    "test": "vitest run",
    "test:watch": "vitest"
  },
  "dependencies": {
    "ripple": "*"
  },
  "packageManager": "pnpm@10.18.2"
}
