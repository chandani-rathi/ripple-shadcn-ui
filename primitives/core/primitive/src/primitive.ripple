import { tracked } from 'ripple';
import type { Tracked } from "ripple";

export function isTrackedObject(v: Tracked) {
    return typeof v === 'object' && v !== null && typeof (v.f) === 'number';
}

export function trackProps(v: Record<string|symbol, any>) {
	var isTracked = isTrackedObject(v);

	if (isTracked || typeof v !== 'object' || v === null || Array.isArray(v)) {
		throw new TypeError('Invalid value: expected a non-tracked object');
	}

	var descriptors = Object.getOwnPropertyDescriptors(v);

	return new Proxy(v, {
		get(target, key) {
			if (isTrackedObject(target[key])) {
				return target[key];
			}
			let t = tracked(undefined);
			const exists = !!descriptors[key];
			if (exists) {
				t = Object.defineProperty(t, '__v', descriptors[key]);
			}
			return t;
		}
	});
}