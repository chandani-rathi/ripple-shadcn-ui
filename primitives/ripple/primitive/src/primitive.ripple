import { track } from "ripple"
import type { PropsWithChildren } from "ripple";
import { normalizeStyleObject } from "@ripple-primitives/utils";
import type { Tracked } from "ripple";

const NODES = [
    'a',
    'abbr',
    'address',
    'article',
    'aside',
    'audio',
    'button',
    'canvas',
    'cite',
    'code',
    'data',
    'datalist',
    'dd',
    'del',
    'details',
    'dfn',
    'dialog',
    'div',
    'dl',
    'dt',
    'em',
    'fieldset',
    'figcaption',
    'figure',
    'footer',
    'form',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'header',
    'hr',
    'i',
    'iframe',
    'img',
    'input',
    'ins',
    'kbd',
    'label',
    'legend',
    'li',
    'main',
    'map',
    'mark',
    'menu',
    'meter',
    'nav',
    'ol',
    'output',
    'p',
    'picture',
    'pre',
    'progress',
    'q',
    'rp',
    'rt',
    'ruby',
    's',
    'samp',
    'section',
    'select',
    'small',
    'source',
    'span',
    'strong',
    'sub',
    'summary',
    'sup',
    'table',
    'tbody',
    'td',
    'template',
    'textarea',
    'tfoot',
    'th',
    'thead',
    'time',
    'title',
    'tr',
    'u',
    'ul',
    'video',
    'svg',
] as const;

type Primitives = { [E in (typeof NODES)[number]]: PrimitiveProps<E> };

type PrimitiveProps<T extends object = {}> = PropsWithChildren<{ asChild?: boolean }>

/*
    <Primitive.div />
    <Primitive.button />
*/
export const Primitive = NODES.reduce(
    (primitive, node) => {
        component Node(props : PrimitiveProps){
            const {children, asChild, style, ...nodeProps} = trackProps(props);
            let tag = track(node);
            if(@asChild) {
                <@children style={normalizeStyleObject(@style)} {...nodeProps} />
            }
            else {
                <@tag style={normalizeStyleObject(@style)} {...nodeProps}>
                    <@children />
                </@tag> 
            }
        }
        return {
            ...primitive, [node]: Node
        }
    }, {} as Primitives
)


const CreateProvider = (providerContext) => {
  component Provider({ children, ...props }) {
    let context = providerContext.get();
    context = {...context, ...props }
    providerContext.set(context);
    <@children />
  }

  return Provider;
}

export function isTrackedObject(v: Tracked) {
    return typeof v === 'object' && v !== null && typeof (v.f) === 'number';
}

export function isFunction(v: Tracked) {
    return typeof v === 'function';
}

export function trackProps(v: Record<string|symbol, any>) {
	var isTracked = isTrackedObject(v);

	if (isTracked || typeof v !== 'object' || v === null || Array.isArray(v)) {
		throw new TypeError('Invalid value: expected a non-tracked object');
	}

	var descriptors = Object.getOwnPropertyDescriptors(v);

	return new Proxy(v, {
		get(target, key) {
			if (isTrackedObject(target[key]) || isFunction(target[key])) {
				return target[key];
			}
            
			let t = undefined;
			const exists = !!descriptors[key];
			if (exists) {
                t = track(undefined);
				t = Object.defineProperty(t, '__v', descriptors[key]);
                t.isTrackedProxy = !descriptors[key].set;
			}
			return t;
		}
	});
}