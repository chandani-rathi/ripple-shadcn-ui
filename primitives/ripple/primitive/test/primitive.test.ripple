import * as ripple from 'ripple';
import { Primitive } from "@/primitive.ripple";

function render(component, target) {
    ripple.mount(component, {
		target
	});
}

test("Primitive UI", () => {
    let container = document.createElement('div');
    Object.keys(Primitive).forEach(
        node => {
            render(Primitive[node], container);
            expect(container.children[0]).toMatchSnapshot()
        }
    )
});

test("button click", () => {
    let container = document.createElement('div');
    const onClick = vi.fn();
    component Test(){
        <Primitive.button id="test" data-state="open" onClick={onClick}>
            {"Test"}
        </Primitive.button>
    }
    render(Test, container);
    const button = container.querySelector("#test");
    expect(button.id).toBe("test");
    expect(button).toHaveAttribute('data-state', 'open')
    button.click();
    expect(onClick).toHaveBeenCalled(1);
})