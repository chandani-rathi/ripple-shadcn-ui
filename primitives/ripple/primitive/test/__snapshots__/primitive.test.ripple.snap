// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Primitive UI 1`] = `
<a>
  <!---->
  <!---->
  <!---->
</a>
`;

exports[`Primitive UI 2`] = `
<button>
  <!---->
</button>
`;

exports[`Primitive UI 3`] = `
<div>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`Primitive UI 4`] = `
<form>
  <!---->
  <!---->
  <!---->
</form>
`;

exports[`Primitive UI 5`] = `
<h2>
  <!---->
  <!---->
  <!---->
</h2>
`;

exports[`Primitive UI 6`] = `
<h3>
  <!---->
  <!---->
  <!---->
</h3>
`;

exports[`Primitive UI 7`] = `
<img>
  <!---->
  <!---->
  <!---->
</img>
`;

exports[`Primitive UI 8`] = `
<input>
  <!---->
  <!---->
  <!---->
</input>
`;

exports[`Primitive UI 9`] = `
<label>
  <!---->
  <!---->
  <!---->
</label>
`;

exports[`Primitive UI 10`] = `
<li>
  <!---->
  <!---->
  <!---->
</li>
`;

exports[`Primitive UI 11`] = `
<nav>
  <!---->
  <!---->
  <!---->
</nav>
`;

exports[`Primitive UI 12`] = `
<ol>
  <!---->
  <!---->
  <!---->
</ol>
`;

exports[`Primitive UI 13`] = `
<p>
  <!---->
  <!---->
  <!---->
</p>
`;

exports[`Primitive UI 14`] = `
<select>
  <!---->
  <!---->
  <!---->
</select>
`;

exports[`Primitive UI 15`] = `
<span>
  <!---->
  <!---->
  <!---->
</span>
`;

exports[`Primitive UI 16`] = `
<svg>
  <!---->
  <!---->
  <!---->
</svg>
`;

exports[`Primitive UI 17`] = `
<ul>
  <!---->
  <!---->
  <!---->
</ul>
`;
